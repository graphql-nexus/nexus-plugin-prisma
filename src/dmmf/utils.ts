import * as Photon from '@prisma/photon'
import { transform } from './transformer'
import { DMMFClass } from './DMMFClass'

export function fromPhotonDMMF(photonDMMF: Photon.DMMF.Document): DMMFClass {
  return new DMMFClass(transform(photonDMMF))
}

export const get = (photonClientPackagePath: string): DMMFClass => {
  let photonClientPackage
  try {
    photonClientPackage = require(photonClientPackagePath)
  } catch (error) {
    throw new Error(
      `Could not find photon package at ${photonClientPackagePath}. Check that you have configured your Photon generator block in schema.prisma correctly and run prisma generate.`,
    )
  }
  return fromPhotonDMMF(photonClientPackage.dmmf)
}
