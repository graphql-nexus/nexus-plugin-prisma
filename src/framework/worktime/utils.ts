import * as Prisma from '@prisma/sdk'
import chalk from 'chalk'
import { WorktimeLens } from 'nexus/plugin'
import stripAnsi from 'strip-ansi'
import * as Scaffolders from './scaffolders'

/**
 * Pinned query-engine version. Calculated at build time and based on `@prisma/cli` version
 */
export const PRISMA_QUERY_ENGINE_VERSION: string = require('@prisma/cli/package.json').prisma.version

/**
 * Get the declared generator blocks in the user's PSL file.
 *
 * If Prisma SDK throws a no-models-found error then it is caught, a model is
 * scaffolded, and this function is re-run.
 */
export async function getGenerators(nexus: WorktimeLens, schemaPath: string): Promise<Prisma.Generator[]> {
  return Prisma.getGenerators({
    schemaPath,
    printDownloadProgress: false,
    version: PRISMA_QUERY_ENGINE_VERSION,
  })
    .catch(async (e: Error) => {
      if (isNoModelsDefinedError(e)) {
        await Scaffolders.exampleModelBlock(schemaPath)
        nexus.log.warn(
          `An example model has been scaffolded for you in ${chalk.bold(
            nexus.layout.projectRelative(schemaPath)
          )}`
        )
        return getGenerators(nexus, schemaPath)
      }
      return Promise.reject(e)
    })
    .then(async (generators) => {
      const hasClientGenerator = generators.find((g) => g.options?.generator.provider === 'prisma-client-js')
      if (!hasClientGenerator) {
        await Scaffolders.prismaClientGeneratorBlock(schemaPath)
        nexus.log.warn(
          `A Prisma Client generator block has been scaffolded for you in ${chalk.bold(
            nexus.layout.projectRelative(schemaPath)
          )}`
        )
        // TODO: Generate it programmatically instead for performance reason
        return getGenerators(nexus, schemaPath)
      } else {
        return generators
      }
    })
}

/**
 * Check if the thrown error is of NoModelsDefined type.
 */
function isNoModelsDefinedError(e: Error): boolean {
  return Boolean(stripAnsi(e.message).match(/.*You don't have any models defined.*/))
}
