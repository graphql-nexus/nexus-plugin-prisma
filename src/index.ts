import * as Builder from './builder'

export type NexusPrismaParams = Builder.Options
export const nexusPrismaPlugin = Builder.build
