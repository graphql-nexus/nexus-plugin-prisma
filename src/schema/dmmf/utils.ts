import { DMMF } from '@prisma/client/runtime'

export const getPrismaClientDmmf = (packagePath: string): DMMF.Document => {
  let dmmf: undefined | DMMF.Document = undefined

  try {
    dmmf = require(packagePath).dmmf
  } catch (error) {
    throw new Error(
      `Failed to import prisma client package at ${packagePath}. The following error occured while trying:
        ${error.stack}`
    )
  }

  if (!dmmf) {
    throw new Error(`\
You most likely forgot to initialize the Prisma Client. Please run \`prisma generate\` and try to run it again.
If that does not solve your problem, please open an issue.`)
  }

  return dmmf
}
