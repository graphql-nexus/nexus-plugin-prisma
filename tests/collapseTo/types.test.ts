import { mutationType } from 'nexus'
import { printSchema } from 'graphql'
import { getTestData, defaultDefinitions, mockConsoleLog } from '../__utils'
import { InputsConfig, PrismaInputFieldName } from '../../src/utils'
import { NexusGraphQLSchema } from 'nexus/dist/core'

describe('collapseTo typegen', () => {
  it('works at plugin-level', async () => {
    const { schema, typegen } = await getTestData({
      pluginOptions: {
        inputs: {
          nests: {
            collapseTo: 'create',
          },
        } as InputsConfig,
      },
    })
    expect(printSchema(schema)).toMatchSnapshot('plugin-level-schema')
    expect(typegen).toMatchSnapshot('plugin-level-typegen')
  })
  it('works at resolver-level', async () => {
    const { schema, typegen } = await getTestData({
      definitions: {
        ...defaultDefinitions,
        mutation: mutationType({
          definition(t: any) {
            t.crud.createOneUser({
              inputs: {
                nests: {
                  collapseTo: 'create',
                },
              } as InputsConfig,
            })
            t.crud.createOneNest()
          },
        }),
      },
    })
    expect(printSchema(schema)).toMatchSnapshot('resolver-level-schema')
    expect(typegen).toMatchSnapshot('resolver-level-typegen')
  })
  it('works for queries', async () => {
    const { schema, typegen } = await getTestData({
      pluginOptions: {
        inputs: {
          createdAt: {
            collapseTo: 'equals',
          },
        } as InputsConfig,
      },
    })
    expect(printSchema(schema)).toMatchSnapshot('query-schema')
    expect(typegen).toMatchSnapshot('query-typegen')
  })
  it('avoids collapsing required fields', async () => {
    let schema: NexusGraphQLSchema
    let typegen: string
    const { $output } = await mockConsoleLog(async () => {
      const data = await getTestData({
        pluginOptions: {
          collapseTo: 'name' as PrismaInputFieldName,
        },
      })
      schema = data.schema
      typegen = data.typegen
    })
    expect(printSchema(schema!)).toMatchSnapshot('avoid-required-schema')
    expect(typegen!).toMatchSnapshot('avoid-required-typegen')
    expect($output).toContain('Not collapsing field "name"')
  })
})
