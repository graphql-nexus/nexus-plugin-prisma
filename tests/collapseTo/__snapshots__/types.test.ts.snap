// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collapseTo typegen avoids collapsing required fields: avoid-required-schema 1`] = `
"scalar DateTime

type Mutation {
  createOneUser(data: UserCreateInput!): User!
  createOneNest(data: NestCreateInput!): Nest!
}

type Nest {
  id: Int!
  createdWithBrowser: String!
  name: String!
}

input NestCreateInput {
  name: String!
  createdWithBrowser: String!
  user: UserCreateOneWithoutNestsInput
}

input NestCreateManyWithoutUserInput {
  create: [NestCreateWithoutUserInput!]
  connect: [NestWhereUniqueInput!]
}

input NestCreateWithoutUserInput {
  name: String!
  createdWithBrowser: String!
}

input NestWhereUniqueInput {
  id: Int
}

type Query {
  user(where: UserWhereUniqueInput!): User
  users(where: StringFilter, skip: Int, after: UserWhereUniqueInput, before: UserWhereUniqueInput, first: Int, last: Int): [User!]!
}

input StringFilter {
  equals: String
  not: String
  in: [String!]
  notIn: [String!]
  lt: String
  lte: String
  gt: String
  gte: String
  contains: String
  startsWith: String
  endsWith: String
}

type User {
  id: Int!
  name: String!
  nests(skip: Int, after: NestWhereUniqueInput, before: NestWhereUniqueInput, first: Int, last: Int): [Nest!]!
  createdWithBrowser: String!
}

input UserCreateInput {
  createdAt: DateTime
  name: String!
  createdWithBrowser: String!
  nests: NestCreateManyWithoutUserInput
}

input UserCreateOneWithoutNestsInput {
  create: UserCreateWithoutNestsInput
  connect: UserWhereUniqueInput
}

input UserCreateWithoutNestsInput {
  createdAt: DateTime
  name: String!
  createdWithBrowser: String!
}

input UserWhereUniqueInput {
  id: Int
}
"
`;

exports[`collapseTo typegen avoids collapsing required fields: avoid-required-typegen 1`] = `
"
interface ModelTypes {
  User: prisma.User
  Nest: prisma.Nest
}
  
interface NexusPrismaInputs {
  Query: {
    users: {
  filtering: 'id' | 'createdAt' | 'name' | 'nests' | 'createdWithBrowser' | 'AND' | 'OR' | 'NOT'
  ordering: 'id' | 'createdAt' | 'name' | 'createdWithBrowser'
}
    nests: {
  filtering: 'id' | 'name' | 'createdWithBrowser' | 'AND' | 'OR' | 'NOT' | 'user'
  ordering: 'id' | 'name' | 'createdWithBrowser'
}

  },
    User: {
    nests: {
  filtering: 'id' | 'name' | 'createdWithBrowser' | 'AND' | 'OR' | 'NOT' | 'user'
  ordering: 'id' | 'name' | 'createdWithBrowser'
}

  },  Nest: {


  }
}

interface NexusPrismaTypes {
  Query: {
    user: 'User'
    users: 'User'
    nest: 'Nest'
    nests: 'Nest'

  },
  Mutation: {
    createOneUser: 'User'
    updateOneUser: 'User'
    updateManyUser: 'BatchPayload'
    deleteOneUser: 'User'
    deleteManyUser: 'BatchPayload'
    upsertOneUser: 'User'
    createOneNest: 'Nest'
    updateOneNest: 'Nest'
    updateManyNest: 'BatchPayload'
    deleteOneNest: 'Nest'
    deleteManyNest: 'BatchPayload'
    upsertOneNest: 'Nest'

  },
  User: {
    id: 'Int'
    createdAt: 'DateTime'
    name: 'String'
    nests: 'Nest'
    createdWithBrowser: 'String'

},  Nest: {
    id: 'Int'
    name: 'String'
    createdWithBrowser: 'String'
    user: 'User'

}
}

interface NexusPrismaMethods {
  User: NexusPrismaFields<'User'>
  Nest: NexusPrismaFields<'Nest'>
  Query: NexusPrismaFields<'Query'>
  Mutation: NexusPrismaFields<'Mutation'>
}
  

declare global {
  type NexusPrisma<
    TypeName extends string,
    ModelOrCrud extends 'model' | 'crud'
  > = GetNexusPrisma<TypeName, ModelOrCrud>;

  // Pre-transform inputs
  interface PrismaInputs {
  
    NestWhereInput: prisma.NestWhereInput
    UserWhereInput: prisma.UserWhereInput
    IdCompoundUniqueInput: prisma.IdCompoundUniqueInput
    UserWhereUniqueInput: prisma.UserWhereUniqueInput
    NestWhereUniqueInput: prisma.NestWhereUniqueInput
    NestCreateWithoutUserInput: prisma.NestCreateWithoutUserInput
    NestCreateManyWithoutUserInput: prisma.NestCreateManyWithoutUserInput
    UserCreateInput: prisma.UserCreateInput
    NestUpdateWithoutUserDataInput: prisma.NestUpdateWithoutUserDataInput
    NestUpdateWithWhereUniqueWithoutUserInput: prisma.NestUpdateWithWhereUniqueWithoutUserInput
    NestScalarWhereInput: prisma.NestScalarWhereInput
    NestUpdateManyDataInput: prisma.NestUpdateManyDataInput
    NestUpdateManyWithWhereNestedInput: prisma.NestUpdateManyWithWhereNestedInput
    NestUpsertWithWhereUniqueWithoutUserInput: prisma.NestUpsertWithWhereUniqueWithoutUserInput
    NestUpdateManyWithoutUserInput: prisma.NestUpdateManyWithoutUserInput
    UserUpdateInput: prisma.UserUpdateInput
    UserUpdateManyMutationInput: prisma.UserUpdateManyMutationInput
    UserCreateWithoutNestsInput: prisma.UserCreateWithoutNestsInput
    UserCreateOneWithoutNestsInput: prisma.UserCreateOneWithoutNestsInput
    NestCreateInput: prisma.NestCreateInput
    UserUpdateWithoutNestsDataInput: prisma.UserUpdateWithoutNestsDataInput
    UserUpsertWithoutNestsInput: prisma.UserUpsertWithoutNestsInput
    UserUpdateOneWithoutNestsInput: prisma.UserUpdateOneWithoutNestsInput
    NestUpdateInput: prisma.NestUpdateInput
    NestUpdateManyMutationInput: prisma.NestUpdateManyMutationInput
    IntFilter: prisma.IntFilter
    StringFilter: prisma.StringFilter
    DateTimeFilter: prisma.DateTimeFilter
    NestFilter: prisma.NestFilter
    UserOrderByInput: prisma.UserOrderByInput
    NestOrderByInput: prisma.NestOrderByInput
  }
}
  "
`;

exports[`collapseTo typegen works at plugin-level: plugin-level-schema 1`] = `
"scalar DateTime

input DateTimeFilter {
  equals: DateTime
  not: DateTime
  in: [DateTime!]
  notIn: [DateTime!]
  lt: DateTime
  lte: DateTime
  gt: DateTime
  gte: DateTime
}

input IntFilter {
  equals: Int
  not: Int
  in: [Int!]
  notIn: [Int!]
  lt: Int
  lte: Int
  gt: Int
  gte: Int
}

type Mutation {
  createOneUser(data: UserCreateInput!): User!
  createOneNest(data: NestCreateInput!): Nest!
}

type Nest {
  id: Int!
  createdWithBrowser: String!
  name: String!
}

input NestCreateInput {
  name: String!
  createdWithBrowser: String!
  user: UserCreateOneWithoutNestsInput
}

input NestCreateWithoutUserInput {
  name: String!
  createdWithBrowser: String!
}

input NestFilter {
  every: NestWhereInput
  some: NestWhereInput
  none: NestWhereInput
}

input NestWhereInput {
  id: IntFilter
  name: StringFilter
  createdWithBrowser: StringFilter
  AND: [NestWhereInput!]
  OR: [NestWhereInput!]
  NOT: [NestWhereInput!]
  user: UserWhereInput
}

input NestWhereUniqueInput {
  id: Int
}

type Query {
  user(where: UserWhereUniqueInput!): User
  users(where: UserWhereInput, skip: Int, after: UserWhereUniqueInput, before: UserWhereUniqueInput, first: Int, last: Int): [User!]!
}

input StringFilter {
  equals: String
  not: String
  in: [String!]
  notIn: [String!]
  lt: String
  lte: String
  gt: String
  gte: String
  contains: String
  startsWith: String
  endsWith: String
}

type User {
  id: Int!
  name: String!
  nests(skip: Int, after: NestWhereUniqueInput, before: NestWhereUniqueInput, first: Int, last: Int): [Nest!]!
  createdWithBrowser: String!
}

input UserCreateInput {
  createdAt: DateTime
  name: String!
  createdWithBrowser: String!
  nests: NestCreateWithoutUserInput
}

input UserCreateOneWithoutNestsInput {
  create: UserCreateWithoutNestsInput
  connect: UserWhereUniqueInput
}

input UserCreateWithoutNestsInput {
  createdAt: DateTime
  name: String!
  createdWithBrowser: String!
}

input UserWhereInput {
  id: IntFilter
  createdAt: DateTimeFilter
  name: StringFilter
  nests: NestFilter
  createdWithBrowser: StringFilter
  AND: [UserWhereInput!]
  OR: [UserWhereInput!]
  NOT: [UserWhereInput!]
}

input UserWhereUniqueInput {
  id: Int
}
"
`;

exports[`collapseTo typegen works at plugin-level: plugin-level-typegen 1`] = `
"
interface ModelTypes {
  User: prisma.User
  Nest: prisma.Nest
}
  
interface NexusPrismaInputs {
  Query: {
    users: {
  filtering: 'id' | 'createdAt' | 'name' | 'nests' | 'createdWithBrowser' | 'AND' | 'OR' | 'NOT'
  ordering: 'id' | 'createdAt' | 'name' | 'createdWithBrowser'
}
    nests: {
  filtering: 'id' | 'name' | 'createdWithBrowser' | 'AND' | 'OR' | 'NOT' | 'user'
  ordering: 'id' | 'name' | 'createdWithBrowser'
}

  },
    User: {
    nests: {
  filtering: 'id' | 'name' | 'createdWithBrowser' | 'AND' | 'OR' | 'NOT' | 'user'
  ordering: 'id' | 'name' | 'createdWithBrowser'
}

  },  Nest: {


  }
}

interface NexusPrismaTypes {
  Query: {
    user: 'User'
    users: 'User'
    nest: 'Nest'
    nests: 'Nest'

  },
  Mutation: {
    createOneUser: 'User'
    updateOneUser: 'User'
    updateManyUser: 'BatchPayload'
    deleteOneUser: 'User'
    deleteManyUser: 'BatchPayload'
    upsertOneUser: 'User'
    createOneNest: 'Nest'
    updateOneNest: 'Nest'
    updateManyNest: 'BatchPayload'
    deleteOneNest: 'Nest'
    deleteManyNest: 'BatchPayload'
    upsertOneNest: 'Nest'

  },
  User: {
    id: 'Int'
    createdAt: 'DateTime'
    name: 'String'
    nests: 'Nest'
    createdWithBrowser: 'String'

},  Nest: {
    id: 'Int'
    name: 'String'
    createdWithBrowser: 'String'
    user: 'User'

}
}

interface NexusPrismaMethods {
  User: NexusPrismaFields<'User'>
  Nest: NexusPrismaFields<'Nest'>
  Query: NexusPrismaFields<'Query'>
  Mutation: NexusPrismaFields<'Mutation'>
}
  

declare global {
  type NexusPrisma<
    TypeName extends string,
    ModelOrCrud extends 'model' | 'crud'
  > = GetNexusPrisma<TypeName, ModelOrCrud>;

  // Pre-transform inputs
  interface PrismaInputs {
  
    NestWhereInput: prisma.NestWhereInput
    UserWhereInput: prisma.UserWhereInput
    IdCompoundUniqueInput: prisma.IdCompoundUniqueInput
    UserWhereUniqueInput: prisma.UserWhereUniqueInput
    NestWhereUniqueInput: prisma.NestWhereUniqueInput
    NestCreateWithoutUserInput: prisma.NestCreateWithoutUserInput
    NestCreateManyWithoutUserInput: prisma.NestCreateManyWithoutUserInput
    UserCreateInput: prisma.UserCreateInput
    NestUpdateWithoutUserDataInput: prisma.NestUpdateWithoutUserDataInput
    NestUpdateWithWhereUniqueWithoutUserInput: prisma.NestUpdateWithWhereUniqueWithoutUserInput
    NestScalarWhereInput: prisma.NestScalarWhereInput
    NestUpdateManyDataInput: prisma.NestUpdateManyDataInput
    NestUpdateManyWithWhereNestedInput: prisma.NestUpdateManyWithWhereNestedInput
    NestUpsertWithWhereUniqueWithoutUserInput: prisma.NestUpsertWithWhereUniqueWithoutUserInput
    NestUpdateManyWithoutUserInput: prisma.NestUpdateManyWithoutUserInput
    UserUpdateInput: prisma.UserUpdateInput
    UserUpdateManyMutationInput: prisma.UserUpdateManyMutationInput
    UserCreateWithoutNestsInput: prisma.UserCreateWithoutNestsInput
    UserCreateOneWithoutNestsInput: prisma.UserCreateOneWithoutNestsInput
    NestCreateInput: prisma.NestCreateInput
    UserUpdateWithoutNestsDataInput: prisma.UserUpdateWithoutNestsDataInput
    UserUpsertWithoutNestsInput: prisma.UserUpsertWithoutNestsInput
    UserUpdateOneWithoutNestsInput: prisma.UserUpdateOneWithoutNestsInput
    NestUpdateInput: prisma.NestUpdateInput
    NestUpdateManyMutationInput: prisma.NestUpdateManyMutationInput
    IntFilter: prisma.IntFilter
    StringFilter: prisma.StringFilter
    DateTimeFilter: prisma.DateTimeFilter
    NestFilter: prisma.NestFilter
    UserOrderByInput: prisma.UserOrderByInput
    NestOrderByInput: prisma.NestOrderByInput
  }
}
  "
`;

exports[`collapseTo typegen works at resolver-level: resolver-level-schema 1`] = `
"scalar DateTime

input DateTimeFilter {
  equals: DateTime
  not: DateTime
  in: [DateTime!]
  notIn: [DateTime!]
  lt: DateTime
  lte: DateTime
  gt: DateTime
  gte: DateTime
}

input IntFilter {
  equals: Int
  not: Int
  in: [Int!]
  notIn: [Int!]
  lt: Int
  lte: Int
  gt: Int
  gte: Int
}

type Mutation {
  createOneUser(data: UserCreateCollapseNestsToCreateInput!): User!
  createOneNest(data: NestCreateInput!): Nest!
}

type Nest {
  id: Int!
  createdWithBrowser: String!
  name: String!
}

input NestCreateInput {
  name: String!
  createdWithBrowser: String!
  user: UserCreateOneWithoutNestsInput
}

input NestCreateWithoutUserCollapseNestsToCreateInput {
  name: String!
  createdWithBrowser: String!
}

input NestFilter {
  every: NestWhereInput
  some: NestWhereInput
  none: NestWhereInput
}

input NestWhereInput {
  id: IntFilter
  name: StringFilter
  createdWithBrowser: StringFilter
  AND: [NestWhereInput!]
  OR: [NestWhereInput!]
  NOT: [NestWhereInput!]
  user: UserWhereInput
}

input NestWhereUniqueInput {
  id: Int
}

type Query {
  user(where: UserWhereUniqueInput!): User
  users(where: UserWhereInput, skip: Int, after: UserWhereUniqueInput, before: UserWhereUniqueInput, first: Int, last: Int): [User!]!
}

input StringFilter {
  equals: String
  not: String
  in: [String!]
  notIn: [String!]
  lt: String
  lte: String
  gt: String
  gte: String
  contains: String
  startsWith: String
  endsWith: String
}

type User {
  id: Int!
  name: String!
  nests(skip: Int, after: NestWhereUniqueInput, before: NestWhereUniqueInput, first: Int, last: Int): [Nest!]!
  createdWithBrowser: String!
}

input UserCreateCollapseNestsToCreateInput {
  createdAt: DateTime
  name: String!
  createdWithBrowser: String!
  nests: NestCreateWithoutUserCollapseNestsToCreateInput
}

input UserCreateOneWithoutNestsInput {
  create: UserCreateWithoutNestsInput
  connect: UserWhereUniqueInput
}

input UserCreateWithoutNestsInput {
  createdAt: DateTime
  name: String!
  createdWithBrowser: String!
}

input UserWhereInput {
  id: IntFilter
  createdAt: DateTimeFilter
  name: StringFilter
  nests: NestFilter
  createdWithBrowser: StringFilter
  AND: [UserWhereInput!]
  OR: [UserWhereInput!]
  NOT: [UserWhereInput!]
}

input UserWhereUniqueInput {
  id: Int
}
"
`;

exports[`collapseTo typegen works at resolver-level: resolver-level-typegen 1`] = `
"
interface ModelTypes {
  User: prisma.User
  Nest: prisma.Nest
}
  
interface NexusPrismaInputs {
  Query: {
    users: {
  filtering: 'id' | 'createdAt' | 'name' | 'nests' | 'createdWithBrowser' | 'AND' | 'OR' | 'NOT'
  ordering: 'id' | 'createdAt' | 'name' | 'createdWithBrowser'
}
    nests: {
  filtering: 'id' | 'name' | 'createdWithBrowser' | 'AND' | 'OR' | 'NOT' | 'user'
  ordering: 'id' | 'name' | 'createdWithBrowser'
}

  },
    User: {
    nests: {
  filtering: 'id' | 'name' | 'createdWithBrowser' | 'AND' | 'OR' | 'NOT' | 'user'
  ordering: 'id' | 'name' | 'createdWithBrowser'
}

  },  Nest: {


  }
}

interface NexusPrismaTypes {
  Query: {
    user: 'User'
    users: 'User'
    nest: 'Nest'
    nests: 'Nest'

  },
  Mutation: {
    createOneUser: 'User'
    updateOneUser: 'User'
    updateManyUser: 'BatchPayload'
    deleteOneUser: 'User'
    deleteManyUser: 'BatchPayload'
    upsertOneUser: 'User'
    createOneNest: 'Nest'
    updateOneNest: 'Nest'
    updateManyNest: 'BatchPayload'
    deleteOneNest: 'Nest'
    deleteManyNest: 'BatchPayload'
    upsertOneNest: 'Nest'

  },
  User: {
    id: 'Int'
    createdAt: 'DateTime'
    name: 'String'
    nests: 'Nest'
    createdWithBrowser: 'String'

},  Nest: {
    id: 'Int'
    name: 'String'
    createdWithBrowser: 'String'
    user: 'User'

}
}

interface NexusPrismaMethods {
  User: NexusPrismaFields<'User'>
  Nest: NexusPrismaFields<'Nest'>
  Query: NexusPrismaFields<'Query'>
  Mutation: NexusPrismaFields<'Mutation'>
}
  

declare global {
  type NexusPrisma<
    TypeName extends string,
    ModelOrCrud extends 'model' | 'crud'
  > = GetNexusPrisma<TypeName, ModelOrCrud>;

  // Pre-transform inputs
  interface PrismaInputs {
  
    NestWhereInput: prisma.NestWhereInput
    UserWhereInput: prisma.UserWhereInput
    IdCompoundUniqueInput: prisma.IdCompoundUniqueInput
    UserWhereUniqueInput: prisma.UserWhereUniqueInput
    NestWhereUniqueInput: prisma.NestWhereUniqueInput
    NestCreateWithoutUserInput: prisma.NestCreateWithoutUserInput
    NestCreateManyWithoutUserInput: prisma.NestCreateManyWithoutUserInput
    UserCreateInput: prisma.UserCreateInput
    NestUpdateWithoutUserDataInput: prisma.NestUpdateWithoutUserDataInput
    NestUpdateWithWhereUniqueWithoutUserInput: prisma.NestUpdateWithWhereUniqueWithoutUserInput
    NestScalarWhereInput: prisma.NestScalarWhereInput
    NestUpdateManyDataInput: prisma.NestUpdateManyDataInput
    NestUpdateManyWithWhereNestedInput: prisma.NestUpdateManyWithWhereNestedInput
    NestUpsertWithWhereUniqueWithoutUserInput: prisma.NestUpsertWithWhereUniqueWithoutUserInput
    NestUpdateManyWithoutUserInput: prisma.NestUpdateManyWithoutUserInput
    UserUpdateInput: prisma.UserUpdateInput
    UserUpdateManyMutationInput: prisma.UserUpdateManyMutationInput
    UserCreateWithoutNestsInput: prisma.UserCreateWithoutNestsInput
    UserCreateOneWithoutNestsInput: prisma.UserCreateOneWithoutNestsInput
    NestCreateInput: prisma.NestCreateInput
    UserUpdateWithoutNestsDataInput: prisma.UserUpdateWithoutNestsDataInput
    UserUpsertWithoutNestsInput: prisma.UserUpsertWithoutNestsInput
    UserUpdateOneWithoutNestsInput: prisma.UserUpdateOneWithoutNestsInput
    NestUpdateInput: prisma.NestUpdateInput
    NestUpdateManyMutationInput: prisma.NestUpdateManyMutationInput
    IntFilter: prisma.IntFilter
    StringFilter: prisma.StringFilter
    DateTimeFilter: prisma.DateTimeFilter
    NestFilter: prisma.NestFilter
    UserOrderByInput: prisma.UserOrderByInput
    NestOrderByInput: prisma.NestOrderByInput
  }
}
  "
`;

exports[`collapseTo typegen works for queries: query-schema 1`] = `
"input DateTime

input IntFilter {
  equals: Int
  not: Int
  in: [Int!]
  notIn: [Int!]
  lt: Int
  lte: Int
  gt: Int
  gte: Int
}

type Mutation {
  createOneUser(data: UserCreateInput!): User!
  createOneNest(data: NestCreateInput!): Nest!
}

type Nest {
  id: Int!
  createdWithBrowser: String!
  name: String!
}

input NestCreateInput {
  name: String!
  createdWithBrowser: String!
  user: UserCreateOneWithoutNestsInput
}

input NestCreateManyWithoutUserInput {
  create: [NestCreateWithoutUserInput!]
  connect: [NestWhereUniqueInput!]
}

input NestCreateWithoutUserInput {
  name: String!
  createdWithBrowser: String!
}

input NestFilter {
  every: NestWhereInput
  some: NestWhereInput
  none: NestWhereInput
}

input NestWhereInput {
  id: IntFilter
  name: StringFilter
  createdWithBrowser: StringFilter
  AND: [NestWhereInput!]
  OR: [NestWhereInput!]
  NOT: [NestWhereInput!]
  user: UserWhereInput
}

input NestWhereUniqueInput {
  id: Int
}

type Query {
  user(where: UserWhereUniqueInput!): User
  users(where: UserWhereInput, skip: Int, after: UserWhereUniqueInput, before: UserWhereUniqueInput, first: Int, last: Int): [User!]!
}

input StringFilter {
  equals: String
  not: String
  in: [String!]
  notIn: [String!]
  lt: String
  lte: String
  gt: String
  gte: String
  contains: String
  startsWith: String
  endsWith: String
}

type User {
  id: Int!
  name: String!
  nests(skip: Int, after: NestWhereUniqueInput, before: NestWhereUniqueInput, first: Int, last: Int): [Nest!]!
  createdWithBrowser: String!
}

input UserCreateInput {
  createdAt: DateTime
  name: String!
  createdWithBrowser: String!
  nests: NestCreateManyWithoutUserInput
}

input UserCreateOneWithoutNestsInput {
  create: UserCreateWithoutNestsInput
  connect: UserWhereUniqueInput
}

input UserCreateWithoutNestsInput {
  createdAt: DateTime
  name: String!
  createdWithBrowser: String!
}

input UserWhereInput {
  id: IntFilter
  createdAt: DateTime
  name: StringFilter
  nests: NestFilter
  createdWithBrowser: StringFilter
  AND: [UserWhereInput!]
  OR: [UserWhereInput!]
  NOT: [UserWhereInput!]
}

input UserWhereUniqueInput {
  id: Int
}
"
`;

exports[`collapseTo typegen works for queries: query-typegen 1`] = `
"
interface ModelTypes {
  User: prisma.User
  Nest: prisma.Nest
}
  
interface NexusPrismaInputs {
  Query: {
    users: {
  filtering: 'id' | 'createdAt' | 'name' | 'nests' | 'createdWithBrowser' | 'AND' | 'OR' | 'NOT'
  ordering: 'id' | 'createdAt' | 'name' | 'createdWithBrowser'
}
    nests: {
  filtering: 'id' | 'name' | 'createdWithBrowser' | 'AND' | 'OR' | 'NOT' | 'user'
  ordering: 'id' | 'name' | 'createdWithBrowser'
}

  },
    User: {
    nests: {
  filtering: 'id' | 'name' | 'createdWithBrowser' | 'AND' | 'OR' | 'NOT' | 'user'
  ordering: 'id' | 'name' | 'createdWithBrowser'
}

  },  Nest: {


  }
}

interface NexusPrismaTypes {
  Query: {
    user: 'User'
    users: 'User'
    nest: 'Nest'
    nests: 'Nest'

  },
  Mutation: {
    createOneUser: 'User'
    updateOneUser: 'User'
    updateManyUser: 'BatchPayload'
    deleteOneUser: 'User'
    deleteManyUser: 'BatchPayload'
    upsertOneUser: 'User'
    createOneNest: 'Nest'
    updateOneNest: 'Nest'
    updateManyNest: 'BatchPayload'
    deleteOneNest: 'Nest'
    deleteManyNest: 'BatchPayload'
    upsertOneNest: 'Nest'

  },
  User: {
    id: 'Int'
    createdAt: 'DateTime'
    name: 'String'
    nests: 'Nest'
    createdWithBrowser: 'String'

},  Nest: {
    id: 'Int'
    name: 'String'
    createdWithBrowser: 'String'
    user: 'User'

}
}

interface NexusPrismaMethods {
  User: NexusPrismaFields<'User'>
  Nest: NexusPrismaFields<'Nest'>
  Query: NexusPrismaFields<'Query'>
  Mutation: NexusPrismaFields<'Mutation'>
}
  

declare global {
  type NexusPrisma<
    TypeName extends string,
    ModelOrCrud extends 'model' | 'crud'
  > = GetNexusPrisma<TypeName, ModelOrCrud>;

  // Pre-transform inputs
  interface PrismaInputs {
  
    NestWhereInput: prisma.NestWhereInput
    UserWhereInput: prisma.UserWhereInput
    IdCompoundUniqueInput: prisma.IdCompoundUniqueInput
    UserWhereUniqueInput: prisma.UserWhereUniqueInput
    NestWhereUniqueInput: prisma.NestWhereUniqueInput
    NestCreateWithoutUserInput: prisma.NestCreateWithoutUserInput
    NestCreateManyWithoutUserInput: prisma.NestCreateManyWithoutUserInput
    UserCreateInput: prisma.UserCreateInput
    NestUpdateWithoutUserDataInput: prisma.NestUpdateWithoutUserDataInput
    NestUpdateWithWhereUniqueWithoutUserInput: prisma.NestUpdateWithWhereUniqueWithoutUserInput
    NestScalarWhereInput: prisma.NestScalarWhereInput
    NestUpdateManyDataInput: prisma.NestUpdateManyDataInput
    NestUpdateManyWithWhereNestedInput: prisma.NestUpdateManyWithWhereNestedInput
    NestUpsertWithWhereUniqueWithoutUserInput: prisma.NestUpsertWithWhereUniqueWithoutUserInput
    NestUpdateManyWithoutUserInput: prisma.NestUpdateManyWithoutUserInput
    UserUpdateInput: prisma.UserUpdateInput
    UserUpdateManyMutationInput: prisma.UserUpdateManyMutationInput
    UserCreateWithoutNestsInput: prisma.UserCreateWithoutNestsInput
    UserCreateOneWithoutNestsInput: prisma.UserCreateOneWithoutNestsInput
    NestCreateInput: prisma.NestCreateInput
    UserUpdateWithoutNestsDataInput: prisma.UserUpdateWithoutNestsDataInput
    UserUpsertWithoutNestsInput: prisma.UserUpsertWithoutNestsInput
    UserUpdateOneWithoutNestsInput: prisma.UserUpdateOneWithoutNestsInput
    NestUpdateInput: prisma.NestUpdateInput
    NestUpdateManyMutationInput: prisma.NestUpdateManyMutationInput
    IntFilter: prisma.IntFilter
    StringFilter: prisma.StringFilter
    DateTimeFilter: prisma.DateTimeFilter
    NestFilter: prisma.NestFilter
    UserOrderByInput: prisma.UserOrderByInput
    NestOrderByInput: prisma.NestOrderByInput
  }
}
  "
`;
